[ABI break] Add new structs with version info and readonly flag

tirthasheshpatel opened the pull request:
Closes #34 
Supersedes #72

This PR adds version info to the `DLTensor` struct. This is under discussion in #34 (see [this comment](https://github.com/dmlc/dlpack/issues/34#issuecomment-1078962112)). Since this is an ABI break, the ABI version has been incremented to 2 and the DLPack Version has been bumped to v0.7.0. This also updates the Python spec with a detailed upgrade policy and future ABI compatibility. The upgrade policy has been designed such that the libraries that want to keep supporting (exporting and importing) the old ABI or support multiple DLPack and ABI versions can do so without breaking compatibility with the old Python API.

tirthasheshpatel commented on include/dlpack/dlpack.h:
```diff
@@ -200,6 +209,15 @@ typedef struct {
   int64_t* strides;
   /*! \brief The offset in bytes to the beginning pointer to data */
   uint64_t byte_offset;
+  /*! \brief Mark the data readonly. */
+  uint8_t readonly;
+  /*!
+  * \brief Endianness of the data. 1 for non-native endianness and
+  *  0 for native endianness.
+  */
+  uint8_t endianness;
+  /*! \brief Alignment of the data. */
+  uint32_t alignment;
```
These changes still need to be discussed. Since we are breaking the ABI by adding the version info, it would be good to discuss these here and add them too (so we don't need to break the ABI again)

tqchen commented on include/dlpack/dlpack.h:
```diff
@@ -158,6 +158,15 @@ typedef struct {
+  */
+  struct {
+    uint8_t dlpack;
+    uint8_t abi;
```
Let us define DLPackVersion struct, for backward compatibility, consider put it in the end(after byte_offset), we need to consider the alignment properties of these fields.

tqchen replied:
This will also make followup compiler support easier

tirthasheshpatel replied:
Done, thanks!

tirthasheshpatel commented on docs/source/python_spec.rst:
```diff
@@ -65,24 +66,27 @@ struct members, gray text enum values of supported devices and data
+within ``from_dlpack`` - therefore it is consumed exactly once, and it
+will not be visible to users of the Python API.
+
+The producer must set the ``PyCapsule`` name to ``"dltensor"`` if ABI
+version 1 is requested and ``"versioned_dltensor"`` if ABI version >= 2
```
> I asked in [#98 (review)](https://github.com/dmlc/dlpack/pull/98#pullrequestreview-903401339) and I will ask again: With all these changes flush in, shouldn't we bump the DLPack version at least one more time to include #98 #100 before breaking ABIs? Otherwise, ABI version is never set to 1 in any outside work; it only lives in the dev branch here, and this doc change doesn't really make much sense.

Oh, yes, thanks for noticing! I will propose a PR bumping the DLPack version and rebase here once that's merged.

tirthasheshpatel replied:
I have bumped the DLPack version here itself since it makes sense to upgrade both the DLPack and ABI version for every ABI break. Sounds good?

wjakob commented:
An ABI break may also be an opportunity to reexamine the signed/unsignedness of all `DLTensor` members (PR #102).

tirthasheshpatel commented:
Thanks for the help so far everyone! I think the discussion on #34 has settled on adding the `__dlpack_info__` method which returns the max version supported by the producer. I have updated the PR with the new spec (that includes `__dlpack_info__`). I haven't changed capsule renaming yet. @leofang Let us know if remaining the capsule makes sense to you when the new ABI is exported (as mentioned in #34, it provides some extra protection without breaking backward compatibility).

mattip commented:
I originally was in favor of capsule renaming, but I think it is not needed given the consensus around the protocol. While it does provide another layer of safety, I think it is redundant.

wjakob commented:
There are still a number of things I find confusing about the Python specification. Some of them are related to this specific PR. Some of them are things that are unclear in general.

1. The `stream` argument of `__dlpack__()` can presumably be used to pass a CUDA stream object. Is the interface specific to CUDA? If so, the documentation should probably say so.

CUDA streams are not a natural representation in CUDA. Even in C++, they are simply a pointer to an opaque NVIDIA data structure.  Are they passed as capsules, uintptr_t cast to a python int, etc.? The documented interface lacks type annotations to make this clear.

2. The document mentions a new `__dlpack_info__` function but does not provide a clear type-checkable signature of this new function. This is a recipe for inconsistencies. I recommend providing an example raw python version with MyPy-style type signature as an implementation guide.

3. What are the requirements on the `version` attribute of `__dlpack__`? Is it a Python `int`? If so, this could be stated in the type signature.

tirthasheshpatel commented:
> I originally was in favor of capsule renaming, but I think it is not needed given the consensus around the protocol. While it does provide another layer of safety, I think it is redundant.

Okay, I have removed capsule renaming now.

Also, I have added release notes and updated the DLPack diagram with the new structs. This PR is ready for review again from my side.

tqchen commented:
@tirthasheshpatel just realized that we checkedin the png image to the repo in a previous path. Is it possible to checkin the image to a different location/repo? On one hand we would like to keep repo self-contained, on the other hand, it can be a bit annoying for a repo to contain a rich history of binaries through multiple revisions. 

One approach is to send a PR to a separate repo instead (we used https://github.com/dmlc/web-data) for some of that purposes then link to via https://gitcontent

tqchen commented:
https://github.com/dmlc/dlpack/issues/104

mattip commented on docs/source/python_spec.rst:
```diff
@@ -48,6 +50,14 @@ producer; the producer must synchronize or wait on the stream when necessary.
+doesn't support any version below the producer's maximum version, a
+``BufferError`` should be raised. Similarly, If the producer doesn't
+support the requested version, it should raise a ``BufferError``.
+
```
What happens if the consumer does not specify a version? As I understand things, for the forseeable future the producer should return a V1 structure. So effectively the default is `1`.

tirthasheshpatel replied:
> What happens if the consumer does not specify a version? As I understand things, for the forseeable future the producer should return a V1 structure. So effectively the default is `1`.

Yes, the default is 1. I will update the spec to say that instead.

wjakob commented:
> stream is an integer; the ID of the GPU device where the data is present. It is specific to CUDA and ROCm (and the devices that use a stream mechanism). This is explicitly stated in the [semantics section](https://dmlc.github.io/dlpack/latest/python_spec.html#semantics) of the python specification. Is it not clear enough?

I actually find this part of the documentation quite confusing (That said, I am only familiar with the CUDA and can't speak about ROCm).

There are three concepts in CUDA that play a role here:

- Each compatible graphics card/accelerator in a machine is given *CUDA device* ID.
- An application can talk to such a CUDA device by creating a *CUDA context* ([details](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__CTX.html#group__CUDA__CTX)). The CUDA runtime API creates a default context (_primary context_), but this is not necessarily only one. An application could in principle create multiple contexts to launch applications on the same GPU, each with its own virtual address space on the GPU (meaning that a memory allocation in one context cannot be accessed by another context).
- In each context, the application can create multiple *CUDA streams* ([details](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__STREAM.html#group__CUDA__STREAM)). A CUDA stream has the purpose of ordering kernel launches with respect to each other. It does *not* play a role with regards to the residency of memory allocations. (Although, to make this even more complicated, the latest CUDA versions now also have what is called stream-ordered memory allocations,I  digress...)

I think that there is I think a fundamental confusion between the word "stream" and "CUDA context" or "CUDA device ID" in the DLPack documentation. I suspect that will you want to call this "device ID" and not "stream" and also establish the convention that CUDA memory allocations are assumed to be located in the primary context associated with the associated CUDA device ID.

tqchen commented on include/dlpack/dlpack.h:
```diff
@@ -200,27 +210,60 @@ typedef struct {
+  DLPackVersion version;
+  /*! \brief Mark the data readonly. */
+  uint8_t readonly;
+} DLTensorVersioned;
```
Actually one thing that might be useful is to think about how to minimizing the change. Specifically, in light of S0 in https://github.com/dmlc/dlpack/issues/104. We may not need to introduce a separate versioned struct. 

The old framework can still use the DLTensor(as implemented in the versioned version) as it is, as long as there is no reliance of sizeof(DLTensor). Perhaps we can rename the old one as DLTensorLegacy; In the similar spirit, it is safe to `static_cast`  a DLTensor* to DLTensorLegacy*, as a result, we do not need to introduce two versions of DLManagedTensor

tirthasheshpatel replied:
> The old framework can still use the DLTensor(as implemented in the versioned version) as it is, as long as there is no reliance of sizeof(DLTensor). Perhaps we can rename the old one as DLTensorLegacy; In the similar spirit, it is safe to `static_cast` a DLTensor* to DLTensorLegacy*, as a result, we do not need to introduce two versions of DLManagedTensor

But DLManagedTensor doesn't have a pointer to the DLTensor struct. Instead, it is just a plain `DLTensor dl_tensor` field. So, changing the sizeof(DLTensor) will change alignment and size of DLManagedTensor. Which is why we need to rename both structs. I also considered renaming the old structs to `DLTensorLegacy` and `DLManagedTensorLegacy` but it seems equivalent to instead add new structs with a different name (new implementations will be able to see both and can use the struct they want to export...)

tqchen replied:
Ah sorry, you are indeed right. This would make the change ABI breaking. Although it would make future ABI change hard because all of them will be non-backward compact due to the change of deleter offset. Would be great to have more deliberation of this topic.

One possible choice is to append other flags(version, read_only) to DLManagedTensor instead. This of course won't be too ideal in case we want to change fields in DLTensor, but would maintain backward compact

mattip replied:
Ahh, I see. If we move them to DLManagedTensor, we can ensure the location of the deleter function is still at the same offset. Then the code 
```
DLManagedTensorVersioned *managed = (DLManagedTensorVersioned *)PyCapsule_GetPointer(self, "dltensor");
if (managed->deleter) {
         managed->deleter(managed);
```

tirthasheshpatel replied:
Although not ideal, I too think it'd be better to add the new fields to the DLManagedTensor instead. This would eliminate the requirement of the `*Versioned` structs and also make it easier to implement support for the new ABI.

mattip commented on docs/source/python_spec.rst:
```diff
@@ -16,10 +16,12 @@ The array API will offer the following syntax for data interchange:
+   on the array object, which will be called from within ``from_dlpack``,
+   to access the data, to get the maximum supported DLPack version, and
+   to query what device the array is on (may be needed to pass in the
+   correct stream, e.g. in the case of multiple GPUs).
```
Can you break this into 3 separate points? They can be subpoints of 2

mattip commented on docs/source/python_spec.rst:
```diff
@@ -96,7 +112,7 @@ C API which is called either when the refcount on the capsule named
       PyErr_Fetch(&type, &value, &traceback);
 
-      DLManagedTensor *managed = (DLManagedTensor *)PyCapsule_GetPointer(self, "dltensor");
+      DLManagedTensorVersioned *managed = (DLManagedTensorVersioned *)PyCapsule_GetPointer(self, "dltensor");
```
How can we distinguish between these two cases? For legacy capsules, we must cast to the older struct, right?

tirthasheshpatel replied:
> How can we distinguish between these two cases? For legacy capsules, we must cast to the older struct, right?

The offset of `dl_tensor.device` doesn't change with this ABI breaking change (and we also don't anticipate such a change in the future). So, casting it to either struct and accessing the field should work.

mattip replied:
Ahh, this is code created by the __producer__, not the consumer. When the consumer calls `obj.__dlpack__`, the producer creates a capsule, and the capsule's deleter function consists of the code here.

mattip commented on docs/source/python_spec.rst:
```diff
@@ -136,6 +153,44 @@ refer to `github.com/dmlc/dlpack <https://github.com/dmlc/dlpack>`_.
+  (``DLTensorVersioned`` and ``DLManagedTensorVersioned``) should be returned.
+  If the requested version is not supported, ``__dlpack__`` should raise a
+  ``BufferError``.
+* Producers should implement a ``__dlpack_info__`` method that returns the
+  maximum supported DLPack version.
```
```suggestion
* Producers should implement a ``__dlpack_info__`` method that returns the
  maximum supported DLPack version. If this method does not exist, the consumer
  must use the old ABI.
```

tirthasheshpatel replied:
I added this information in the point below this one (that point is all about what the consumer should do). Does that look good or do you think we should add the point here?

tqchen commented:
closed in favor of #113  

We would like to thanks @tirthasheshpatel for your effort on bringing in this change onward
